File: backend/routes/progress.js
const express = require('express');
const db = require('../db');
const { identify, requireUser } = require('../lib/telegramAuth');
const { progressFromCompleted, categoryProgressPct, TOTAL_TASKS } = require('../lib/leveling');
const { categoryTotals } = require('../lib/taskBank');
const { ACHIEVEMENT_META } = require('../lib/achievements');

const router = express.Router();

router.get('/progress', identify, requireUser, (req, res) => {
  const user = req.user;
  const { progressPct, level } = progressFromCompleted(user.completed_count);

  const doneRows = db.prepare(`
    SELECT category, COUNT(*) as done FROM user_schedule
    WHERE user_id = ? AND status = 'done' GROUP BY category
  `).all(user.id);
  const doneByCategory = {};
  doneRows.forEach(r => { doneByCategory[r.category] = r.done; });

  const totals = user.gender ? categoryTotals(user.gender) : {};
  const categories = Object.keys(totals).map(cat => ({
    name: cat,
    pct: categoryProgressPct(doneByCategory[cat] || 0, totals[cat]),
    done: doneByCategory[cat] || 0,
    total: totals[cat]
  }));

  res.json({
    level, progressPct,
    xp: user.xp,
    completedCount: user.completed_count,
    totalTasks: TOTAL_TASKS,
    streakCurrent: user.streak_current,
    streakBest: user.streak_best,
    categories
  });
});

router.get('/achievements', identify, requireUser, (req, res) => {
  const rows = db.prepare('SELECT code, unlocked_at FROM achievements WHERE user_id = ?').all(req.user.id);
  const unlockedCodes = new Set(rows.map(r => r.code));
  const all = Object.entries(ACHIEVEMENT_META).map(([code, meta]) => ({
    code, ...meta, unlocked: unlockedCodes.has(code)
  }));
  res.json({ achievements: all });
});

module.exports = router;

File: internal/handlers/progress.go
package handlers

import (
	"net/http"

	"version20/internal/achievements"
	"version20/internal/leveling"
)

type CategoryProgress struct {
	Name  string
	Pct   int
	Done  int
	Total int
}

type ProgressData struct {
	Level          int
	XP             int
	CompletedCount int
	TotalTasks     int
	StreakCurrent  int
	StreakBest     int
	Categories     []CategoryProgress
}

func (a *App) handleProgress(w http.ResponseWriter, r *http.Request) {
	user := userFromCtx(r)
	progressPct, level := leveling.ProgressFromCompleted(user.CompletedCount)
	_ = progressPct

	doneRows, err := a.Store.DoneCountsByCategory(user.ID)
	if err != nil {
		a.serverError(w, err)
		return
	}
	doneByCategory := map[string]int{}
	for _, d := range doneRows {
		doneByCategory[d.Category] = d.Done
	}

	totals := a.CategoryTotals(user.Gender.String)
	var categories []CategoryProgress
	for cat, total := range totals {
		categories = append(categories, CategoryProgress{
			Name: cat, Total: total, Done: doneByCategory[cat],
			Pct: leveling.CategoryProgressPct(doneByCategory[cat], total),
		})
	}

	a.render(w, "progress.html", ProgressData{
		Level: level, XP: user.XP, CompletedCount: user.CompletedCount,
		TotalTasks: leveling.TotalTasks, StreakCurrent: user.StreakCurrent,
		StreakBest: user.StreakBest, Categories: categories,
	})
}

type AchievementItem struct {
	achievements.Meta
	Unlocked bool
}

func (a *App) handleAchievements(w http.ResponseWriter, r *http.Request) {
	user := userFromCtx(r)
	rows, err := a.Store.UserAchievements(user.ID)
	if err != nil {
		a.serverError(w, err)
		return
	}
	unlockedSet := map[string]bool{}
	for _, row := range rows {
		unlockedSet[row.Code] = true
	}

	items := make([]AchievementItem, 0, len(achievements.Metas))
	for _, m := range achievements.Metas {
		items = append(items, AchievementItem{Meta: m, Unlocked: unlockedSet[m.Code]})
	}

	a.render(w, "achievements.html", items)
}
